### PR TITLE
NoImplicityAny for Snapshot and Publisher file in SRC folder

### PR DESCRIPTION
**Pull request by praveenqlogic**
Fixes #<issue_number_goes_here> (it's a good idea to open an issue first for discussion)

- [ ] Tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -33,14 +33,95 @@ const PKG = require('../../package.json');
...
+  (err: Error|null, res?: boolean): void;
+}
+
+export type GetCallOptions = {
+  autoCreate?: boolean
+};
+
+export interface PushConfig {
+  pushEndpoint: string;
+  attibutes?: Map<string, string>;
```
Comment (romlogic) [inline src/index.ts]: attributes?

File: proto/pubsub.d.ts [hunk 2]
```diff
@@ -0,0 +1,8183 @@
+import * as $protobuf from "protobufjs";
```
Comment (romlogic) [inline proto/pubsub.d.ts]: is this a new file? copyright statements?

File: proto/pubsub.d.ts [hunk 3]
```diff
@@ -0,0 +1,8183 @@
...
+/** Namespace google. */
+export namespace google {
+
+    /** Namespace pubsub. */
+    namespace pubsub {
+
+        /** Namespace v1. */
+        namespace v1 {
+
+            /** Represents a Publisher */
```
Comment (romlogic) [inline proto/pubsub.d.ts]: full stop after the sentence.

File: src/index.ts (hunk 1 above)
Comment (praveenqlogic) [reply src/index.ts]: done

File: proto/pubsub.d.ts (hunk 2 above)
Comment (praveenqlogic) [reply proto/pubsub.d.ts]: system generated file!

File: proto/pubsub.d.ts (hunk 3 above)
Comment (praveenqlogic) [reply proto/pubsub.d.ts]: system generated file

File: src/publisher.ts [hunk 4]
```diff
@@ -91,7 +92,7 @@ export class Publisher {
     };
     this.settings = {
       batching: {
-        maxBytes: Math.min(options.batching.maxBytes, Math.pow(1024, 2) * 9),
+        maxBytes: Math.min(options.batching.maxBytes!, Math.pow(1024, 2) * 9),
```
Comment (joecervino) [inline src/publisher.ts]: Just wondering why you added the `!`

File: src/snapshot.ts [hunk 5]
```diff
@@ -185,7 +188,10 @@ export class Snapshot {
    *   const apiResponse = data[0];
    * });
    */
-  delete(callback) {
+  delete(): Promise<google.protobuf.Empty>;
```
Comment (joecervino) [inline src/snapshot.ts:123]: Why is this needed?

File: src/publisher.ts [hunk 6]
```diff
@@ -237,13 +246,17 @@ export class Publisher {
    * @param {object} attributes The message attributes.
    * @param {function} callback The callback function.
    */
-  queue_(data, attrs, callback) {
-    this.inventory_.queued.push({
+  queue_(data: Buffer, attrs: Attributes): Promise<string>;
```
Comment (joecervino) [inline src/publisher.ts:285]: Maybe we can keep the attibutes param name consistent (publish has them down as attributes)

**Review by joecervino — COMMENTED** [review]
Overall looks good :+1:

File: src/publisher.ts (hunk 4 above)
Comment (praveenqlogic) [reply src/publisher.ts]: here `!` is handling variable undefined.

File: src/snapshot.ts (hunk 5 above)
Comment (praveenqlogic) [reply src/snapshot.ts:123]: function signature for information of overloading!

File: src/publisher.ts (hunk 6 above)
Comment (praveenqlogic) [reply src/publisher.ts:285]: 👍

File: src/index.ts (hunk 1 above)
Comment (AVaksman) [reply src/index.ts]: Note, we need only one `PushConfig` interface, this one or the one in `subscription.ts` [module](https://github.com/praveenqlogic/nodejs-pubsub/blob/master/src/subscription.ts#L53). 

Note2. this `PushConfig` is being imported in `subscription.ts` module.

File: src/index.ts [hunk 7]
```diff
@@ -44,6 +44,11 @@ const opts = {} as gax.GrpcClientOptions;
 const {grpc} = new gax.GrpcClient(opts);
 
 
+export interface SnapshotParent extends PubSub {
```
Comment (AVaksman) [inline src/index.ts]: This interface is not needed

`shapshot`'s parent should be `Subscription|PubSub`

File: src/index.ts [hunk 8]
```diff
@@ -61,23 +66,48 @@ export interface PushConfig {
   attibutes?: Map<string, string>;
 }
 
+export interface Inventory {
+  callbacks?: Array<RequestCallback<string>>;
+  queued?: Array<{}>;
+  bytes: number;
+  ack?: string[];
+  lease?: string[];
+  nack?: Array<[string, number]>;
```
Comment (AVaksman) [inline src/index.ts]: where `ack`, `lease` and `nack` are from?
`inventory_` defenition is [here](https://github.com/praveenqlogic/nodejs-pubsub/pull/2/files#diff-2a9261ca5959530e957cf7bfbbab8ea9L99)

File: src/index.ts [hunk 9]
```diff
@@ -61,23 +66,48 @@ export interface PushConfig {
...
+}
+
+export interface FlowControl {
```
Comment (AVaksman) [inline src/index.ts]: can you use [`FlowControlOptions`](https://github.com/praveenqlogic/nodejs-pubsub/blob/master/src/lease-manager.ts#L41) instead of this interface?

Also please use same interface [here](https://github.com/praveenqlogic/nodejs-pubsub/pull/2/files#diff-f41e9d04a45c83f3b6f6e630f10117feR197)

**Comment by AVaksman** [comment]
use `this.request<google.pubsub.v1.Subscription>(` [here](https://github.com/praveenqlogic/nodejs-pubsub/blob/c4b28f66c769ad12f49e840afa367f7182e05682/src/index.ts#L438)

File: src/publisher.ts [hunk 10]
```diff
@@ -157,17 +158,20 @@ export class Publisher {
...
    * publisher.publish(data).then((messageId) => {});
    */
-  publish(data: Buffer, attributes?: object): Promise<string>;
+  publish(data: Buffer, attributes?: Attributes): Promise<string>;
   publish(data: Buffer, callback: PublishCallback): void;
-  publish(data: Buffer, attributes: object, callback: PublishCallback): void;
-  publish(data: Buffer, attributes?, callback?): Promise<string>|void {
+  publish(data: Buffer, attributes: Attributes, callback: PublishCallback):
+      void;
+  publish(data: Buffer, attributesOrCallback?, callback?): Promise<string>|
```
Comment (AVaksman) [inline src/publisher.ts]: types for `attributesOrCallback` and `callback`

File: src/publisher.ts [hunk 11]
```diff
@@ -23,6 +23,7 @@ const each = require('async-each');
 import * as extend from 'extend';
 import * as is from 'is';
 import {Topic} from './topic';
+import {Inventory, RequestCallback, Attributes} from '.';
```
Comment (AVaksman) [inline src/publisher.ts]: also import `import {google} from '../proto/pubsub';`

File: src/publisher.ts [hunk 12]
```diff
@@ -247,7 +251,7 @@ export class Publisher {
       topic: this.topic.name,
       messages,
     };
-    this.topic.request(
+    this.topic.request<PublishApiResponse>(
```
Comment (AVaksman) [inline src/publisher.ts]: use `google.pubsub.v1.PublishResponse` type here and I do not think `PublishApiResponse` type is needed anymore.

File: src/snapshot.ts [hunk 13]
```diff
@@ -15,9 +15,12 @@
  */
 
 import {promisifyAll} from '@google-cloud/promisify';
+import {Metadata} from 'grpc';
```
Comment (AVaksman) [inline src/snapshot.ts]: No need of `Metadata` import

File: src/snapshot.ts [hunk 14]
```diff
@@ -15,9 +15,12 @@
...
 import * as is from 'is';
 
-import {PubSub} from '.';
+import {google} from '../proto/pubsub';
+
+import {RequestCallback, SnapshotParent} from '.';
```
Comment (AVaksman) [inline src/snapshot.ts]: some more imports here
```javascript
import {CreateSnapshotCallback, CreateSnapshotResponse, RequestCallback, Subscription} from '.';
import {CallOptions} from 'google-gax';
import { PubSub } from './index';
```

File: src/snapshot.ts [hunk 15]
```diff
@@ -85,14 +88,14 @@ import * as util from './util';
  * });
  */
 export class Snapshot {
-  parent;
+  parent: SnapshotParent;
```
Comment (AVaksman) [inline src/snapshot.ts]: should be `Subscription|PubSub` type

File: src/snapshot.ts [hunk 16]
```diff
@@ -85,14 +88,14 @@ import * as util from './util';
...
   name: string;
   // tslint:disable-next-line variable-name
   Promise?: PromiseConstructor;
-  create;
-  seek;
-  metadata;
-  constructor(parent, name: string) {
+  create!: Function;
+  seek!: Function;
```
Comment (AVaksman) [inline src/snapshot.ts]: suggest to remove these and rather refactor to a permanent methods with nice descriptive error.

File: src/snapshot.ts [hunk 17]
```diff
@@ -85,14 +88,14 @@ import * as util from './util';
...
+  metadata!: Metadata;
```
Comment (AVaksman) [inline src/snapshot.ts]: metadata should be `google.pubsub.v1.Snapshot` type

File: src/snapshot.ts [hunk 18]
```diff
@@ -85,14 +88,14 @@ import * as util from './util';
...
+  constructor(parent: SnapshotParent, name: string) {
```
Comment (AVaksman) [inline src/snapshot.ts]: `Subscription|Pubsub`

File: src/snapshot.ts [hunk 19]
```diff
@@ -85,14 +88,14 @@ import * as util from './util';
...
     if (parent.Promise) {
```
Comment (AVaksman) [inline src/snapshot.ts]: Verify that parent is `instanceof` `PubSub` here

File: src/snapshot.ts [hunk 20]
```diff
@@ -134,7 +137,7 @@ export class Snapshot {
        *   const apiResponse = data[1];
        * });
        */
-      this.create = parent.createSnapshot.bind(parent, name);
+      this.create = parent.createSnapshot!.bind(parent, name);
```
Comment (AVaksman) [inline src/snapshot.ts]: As I suggested above, remove this from the constructor and define as a method, something like this
```javascript
/**
documentation goes here
*/

create(gaxOpts?: CallOptions): Promise<CreateSnapshotResponse>;
  create(callback: CreateSnapshotCallback): void;
  create(gaxOpts: CallOptions, callback: CreateSnapshotCallback): void;
  create(gaxOpts?: CallOptions|CreateSnapshotCallback, callback?: CreateSnapshotCallback): void|Promise<CreateSnapshotResponse> {
    if (!(this.parent instanceof Subscription)) {
      throw new Error(`This is only available if you accessed this object through {@link Subscription#snapshot}`);
    }
    return (this.parent as Subscription).createSnapshot(this.name, gaxOpts! as CallOptions, callback!);
  }
```

File: src/snapshot.ts [hunk 21]
```diff
@@ -162,7 +165,7 @@ export class Snapshot {
        *   const apiResponse = data[0];
        * });
        */
-      this.seek = parent.seek.bind(parent, name);
+      this.seek = parent.seek!.bind(parent, name);
```
Comment (AVaksman) [inline src/snapshot.ts]: same as with `this.create` above

**Review by AVaksman — CHANGES_REQUESTED** [review]

**Comment by AVaksman** [comment]
in `subscription.ts` fix seek() method types to 
```javascript
  seek(snapshot: string|Date, gaxOpts?: CallOptions): Promise<google.pubsub.v1.SeekResponse>;
  seek(snapshot: string|Date, callback: google.pubsub.v1.Subscriber.SeekCallback): void;
  seek(snapshot: string|Date, gaxOpts: CallOptions, callback: google.pubsub.v1.Subscriber.SeekCallback): void;
  seek(
      snapshot: string|Date,
      gaxOptsOrCallback?: CallOptions|google.pubsub.v1.Subscriber.SeekCallback,
      callback?: google.pubsub.v1.Subscriber.SeekCallback):
      void|Promise<google.pubsub.v1.SeekResponse> {
```

**Comment by AVaksman** [comment]
You will need to fix a couple of tests

File: src/index.ts (hunk 8 above)
Comment (praveenqlogic) [reply src/index.ts]: Done! implementation was done for old code, might be got removed after merge with google master

File: src/index.ts (hunk 9 above)
Comment (praveenqlogic) [reply src/index.ts]: done!, newly type added by google master merge after my changes!

File: src/publisher.ts (hunk 10 above)
Comment (praveenqlogic) [reply src/publisher.ts]: done! due to merge changes!

File: src/publisher.ts (hunk 11 above)
Comment (praveenqlogic) [reply src/publisher.ts]: done

File: src/publisher.ts (hunk 12 above)
Comment (praveenqlogic) [reply src/publisher.ts]: Done!

File: src/snapshot.ts (hunk 13 above)
Comment (praveenqlogic) [reply src/snapshot.ts]: removed

**Comment by AVaksman** [comment]
`snapshot` in `subscription.ts` should be 
```javascript
return new Snapshot(this, name);
```

**Comment by AVaksman** [comment]
Here is how you change the tests to pass:
**Part 1**
Change [this](https://github.com/praveenqlogic/nodejs-pubsub/blob/a3b76959b9c3cdc970b098b4cb28df4dc2a9e0fb/test/snapshot.ts#L119-L139)
https://github.com/praveenqlogic/nodejs-pubsub/blob/a3b76959b9c3cdc970b098b4cb28df4dc2a9e0fb/test/snapshot.ts#L119-L139
To this:
import PubSub and Subscription =>`import { PubSub, Subscription } from '../src';`
then
```javascript
    describe('with Subscription parent', () => {
      let pubsub: PubSub;
      let subscription: Subscription;
      before(() => {
        pubsub = new PubSub({projectId: PROJECT_ID});
        subscription = pubsub.subscription('test');
      })
      it('should include the create method', done => {
        sandbox.stub(subscription, 'createSnapshot').callsFake((name: String) => {
          assert.strictEqual(name, FULL_SNAPSHOT_NAME);
          done();
        });
        
        const snapshot = new Snapshot(subscription, SNAPSHOT_NAME);
        snapshot.create(done);
      });

      it('should call the seek method', done => {
        sandbox.stub(subscription, 'seek').callsFake((snapshot) => {
          assert.strictEqual(snapshot, FULL_SNAPSHOT_NAME);
          done()
        })
        const snapshot = new Snapshot(subscription, SNAPSHOT_NAME);
        snapshot.seek(done);
      });
    });
```

**Comment by AVaksman** [comment]
**Part 2**
Change [this](https://github.com/praveenqlogic/nodejs-pubsub/blob/a3b76959b9c3cdc970b098b4cb28df4dc2a9e0fb/test/snapshot.ts#L119-L139)
To this:
```javascript
    describe('with PubSub parent', () => {
      let snapshot;

      beforeEach(() => {
        snapshot = new Snapshot(PUBSUB, SNAPSHOT_NAME);
      });

      it('should throw on create method', () => {
        assert.throws(() => snapshot.create(), /This is only available if you accessed this object through {@link Subscription#snapshot}/);
      });

      it('should throw on seek method', () => {
        assert.throws(() => snapshot.seek(), /This is only available if you accessed this object through {@link Subscription#snapshot}/);
      });
    });
```

**Comment by AVaksman** [comment]
**Part 3**
Change [this](https://github.com/praveenqlogic/nodejs-pubsub/blob/a3b76959b9c3cdc970b098b4cb28df4dc2a9e0fb/test/subscription.ts#L868-L880)
https://github.com/praveenqlogic/nodejs-pubsub/blob/a3b76959b9c3cdc970b098b4cb28df4dc2a9e0fb/test/subscription.ts#L868-L880
To this:
```javascript
  describe('snapshot', () => {
    const SNAPSHOT_NAME = 'a';

    it('should pass this as Subscription', done => {
      const snapshot = subscription.snapshot(SNAPSHOT_NAME);
      assert.ok(snapshot.calledWith_[0] instanceof Subscription);
      done();
    });
  });
```

File: synth.metadata [hunk 22]
```diff
@@ -1,19 +1,19 @@
 {
```
Comment (ajaaym) [inline synth.metadata]: why your PR contains this file?

**Review by ajaaym — CHANGES_REQUESTED** [review]

File: synth.metadata (hunk 22 above)
Comment (praveenqlogic) [reply synth.metadata]: it was from merging from master, i cleaned it.

**Review by AVaksman — APPROVED** [review]

**Comment by praveenqlogic** [comment]
Completed
